File: wg_sharedavatar_cake_expt/MacCore/AvatarCore/bin/jsTests/video_writer.js
/*
Subscriber Test
This example shows how to:
- Initialize avatarCore SDK
- Create a video resource 
- Periodically update the image contents every 30ms
- Destroy the SDK when pressing ctrl+c

Author: yamen@avatarin.com
*/

const CoreSDK = require("../CoreJS/CoreAPI");
const { CoreVideoWrapper } = require("../CoreJS/CoreVideoWrapper");

//setup input library
const readline = require('readline');
readline.emitKeypressEvents(process.stdin);
process.stdin.setRawMode(true);


//Initialize Core
CoreSDK.InitAPI(master = false, appPath = "../")

//Create a video wrapper object, this will handle image resources
let videoWrapper = new CoreVideoWrapper("local-image", (frame) => { });


let stopped = false
let width = 640
let height = 480
let index = 0

//create a buffer holder for an RGB image data (3 channels)
let videoData = new Uint8Array(width * height * 3)
setInterval(() => {
    if (stopped) return;
    //generate a random image
    for (let x = 0; x < width; x++) {
        for (let y = 0; y < height; y++) {
            let index = y * height + x
            videoData[index * 3 + 0] = Math.random() * 255;
            videoData[index * 3 + 1] = Math.random() * 255;
            videoData[index * 3 + 2] = Math.random() * 255;
        }
    }
    //write video frame resource
    console.log(index, " - Sending Frame [", videoWrapper.resource, "]: ", width, "x", height)
    index++;
    videoWrapper.WriteFrame(width, height, CoreSDK.VideoFormat.RGB, videoData)
}, 30)

const ExitApp = () => {
    //destroy and clean up the core
    videoWrapper.Stop();
    videoWrapper.Destroy();
    videoWrapper = undefined
    //destroy and clean up the core
    CoreSDK.DestroyAPI()
    console.log("Core Destroyed!")
    process.exit(0)
};


process.stdin.on('keypress', (str, key) => {
    if (key.ctrl && key.name === 'c') {
        ExitApp()
    }
});

File: wg_sharedavatar_cake_expt/MacCore/AvatarCore/bin/pyTests/video_writer.py
# This Python file uses the following encoding: utf-8
from CorePython.CoreAPI import CoreAPI
from CorePython.CoreVideo import CoreVideo
import threading
import random

import time

core = CoreAPI()
video = None
stopped = False
width = 640
height = 480


def generate_video_data():
    print("Video data")
    if stopped is True:
        return

    video_data = [0 for i in range(width * height * 3)]

    for i in range(0, width):
        for j in range(0, height):
            index = (j * width + i) * 3
            video_data[index + 0] = random.randint(0, 255)
            video_data[index + 1] = random.randint(0, 255)
            video_data[index + 2] = random.randint(0, 255)

    video.write_frame(width, height, video.VideoFormat.RGB, video_data)
    threading.Timer(5, generate_video_data).start()


def stop():
    global stopped
    stopped = True
    global video
    video.destroy()
    video = None
    global core
    core.destroy_api()
    core = None
    print('Core destroyed')


if __name__ == "__main__":

    core.init_api(app_path='./')
    video = CoreVideo('local-image')
    video.start()

    generate_video_data()

    try:
        while True:
            time.sleep(1)
            pass
    except KeyboardInterrupt:
        stop()
